### DISCUSSION TITLE
Typed property must not be accessed before initialization on rules method during wire:model update

### DISCUSSION THREAD

**Discussion by imtiazPabel** [livewire/livewire#3102] [7. Bugs (archived)]
##### Here is the component

```
<?php

namespace App\Http\Livewire\Patient;

use App\Models\Patient;
use Livewire\Component;

class Create extends Component
{
    public Patient $patient;

    public function rules()
    {
        return [
            'patient.number' => "required|unique:App\Models\Patient,number,{$this->patient->id}",
        ];
    }

    public function render()
    {
        return view('livewire.patient.create');
    }
}
```

##### Simple blade page with wire:model
```
<input type="text" wire:model="patient.number" />
```

##### During patient number update, exactly the error I am getting
```
Typed property App\Http\Livewire\Patient\Create::$patient must not be accessed before initialization
```

**Context**
- Livewire version: [2.3.2]
- Laravel version: [8.15.0]
- Browser: [Chrome]

**Comment by abrardev99** [comment]
I think this is not a bug, this is a behavior. 
This should be the fix. 

```
public mount(){
         $patient = new Patient()
}
```

**Comment by imtiazPabel** [comment]
@abrardev99 you should be right to define a new instance on the mount method, if your component purpose is creating a patient. 

I am editing my patient where patient model is binding automatically by livewire without mount method. [Reference](https://laravel-livewire.com/docs/2.x/properties#binding-models)

For editing purpose on component, sometimes you need to access model property on rules like unique validation except id.

Personally, I think for many other reason you need to access model property on rules.

**Comment by 3alampro** [comment]
> public mount(){
>          $patient = new Patient()
> }

I'm having the same issue on both when trying to create new model or editing existing one 

even when initiating the user model on the mount method produce the same error

**Comment by 3alampro** [comment]
the same issue happen when trying to access any typed properties 

```php
public string $password;
```

and try to validate the field under the rules method 

```php
'password' => [
    'required',
    'min:8',
],
```

when the form submit and trying to validate the password input you will get the same error

```
Error
Typed property App\Http\Livewire\User\Form::$password must not be accessed before initialization
```

to fix it we can assign empty string when defining the property the same goes for other types except typed models

**Comment by austenc** [comment]
@kingpabel could you please provide us some more context? Just tested this in one of my own projects and it worked correctly when using route model binding on a full page livewire component. 

Is this a full page component, or are you including it in another page with `@livewire` or `<livewire:`?
What is your route definition? 

Closing, but feel free to reopen if you'd like to present more info and an example repository we may use to reproduce this. 

Thank you!

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by mattsims** [comment]
Hi,

This still appears to be an issue in Livewire v2.4.4 running on PHP v8.

Using model binding in a nested component, then utilising the model ID property in the `rules()` method, results in:

```
Typed property App\Http\Livewire\Projects\Edit::$project must not be accessed before initialization
```

In this case the model property is an existing model passed in to the component from the view like so:

```
@livewire('projects.edit', ['project' => $project])
```

and set within the `mount()` method using the `fill()` helper:

```
public function mount(Project $project)
{
    $this->fill([
        'project' => $project,
    ]);
}
```

The rule in question is:

```
public function rules()
{
    return [
        'project.number' => [
            'required',
            'numeric',
            Rule::unique('projects', 'number')
                ->ignore($this->project->id),
        ],
    ];
}
```

The `$project` property is `null` when dumped out from within the `rules()` method on re-hydration.
